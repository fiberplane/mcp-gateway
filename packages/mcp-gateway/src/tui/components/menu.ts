import { getActiveSessions, getClientInfo } from "../../capture.js";
import type { Registry } from "../../registry.js";
import type { LogEntry } from "../state.js";
import packageJson from "../../../package.json" with { type: "json" };
import {
  CLEAR_SCREEN,
  CYAN,
  DIM,
  formatRelativeTime,
  GRAY,
  GREEN,
  RED,
  RESET_COLOR,
  YELLOW,
} from "./formatting.js";

// Render the main menu
export function renderMenu(registry: Registry, logs: LogEntry[]): string {
  let output = "";

  const activeSessions = getActiveSessions();
  output += `${CYAN}Fiberplane MCP Gateway v${packageJson.version}${RESET_COLOR}\n`;
  output += `${DIM}Gateway: http://localhost:3333${RESET_COLOR}\n`;
  output += `${DIM}MCP: http://localhost:3333/mcp${RESET_COLOR}\n`;
  if (activeSessions.length > 0) {
    output += `${DIM}Active sessions: ${activeSessions.length}${RESET_COLOR}\n`;
  }
  output += "\n";

  if (registry.servers.length === 0) {
    output += `${DIM}No servers registered${RESET_COLOR}\n`;
  } else {
    output += `${CYAN}Servers:${RESET_COLOR}\n`;
    for (const server of registry.servers) {
      const activity = formatRelativeTime(server.lastActivity);
      const encodedName = encodeURIComponent(server.name);
      const proxyUrl = `http://localhost:3333/${encodedName}/mcp`;
      output += "\n";
      output += `${GREEN}${server.name}${RESET_COLOR}\n`;
      output += `${CYAN}${proxyUrl}${RESET_COLOR} ${DIM}→ ${server.url}${RESET_COLOR}\n`;
      output += `${DIM}Last active: ${activity} • ${server.exchangeCount} exchanges${RESET_COLOR}\n`;
    }
  }

  output += "\n";
  output += `${YELLOW}[a]${RESET_COLOR} Add server\n`;

  if (registry.servers.length > 0) {
    output += `${YELLOW}[d]${RESET_COLOR} Delete server\n`;
  } else {
    output += `${DIM}[d] Delete server${RESET_COLOR}\n`;
  }

  if (logs.length > 0) {
    output += `${YELLOW}[c]${RESET_COLOR} Clear activity\n`;
  } else {
    output += `${DIM}[c] Clear activity${RESET_COLOR}\n`;
  }

  output += `${YELLOW}[m]${RESET_COLOR} MCP instructions\n`;
  output += `${YELLOW}[q]${RESET_COLOR} Quit\n`;
  output += "\n";

  // Display recent logs
  if (logs.length > 0) {
    output += `${CYAN}Recent Activity:${RESET_COLOR}\n`;
    for (const log of logs) {
      // Get client info for this session
      const clientInfo = getClientInfo(log.sessionId);
      const clientLabel = clientInfo
        ? `${clientInfo.name}@${clientInfo.version}`
        : "client";

      // Session ID (first 8 chars)
      const sessionIdShort = `[${log.sessionId.slice(0, 8)}]`;

      if (log.direction === "request") {
        // Client → Gateway → Server (request flow)
        const methodDetails = formatRequestDetails(log);
        output += `${DIM}${log.timestamp.slice(11, 19)}${RESET_COLOR} ${GRAY}${sessionIdShort}${RESET_COLOR} ${CYAN}${clientLabel}${RESET_COLOR} → ${log.serverName} ${DIM}${log.method}${methodDetails}${RESET_COLOR}\n`;
      } else {
        // Server → Gateway → Client (response flow)
        const statusColor =
          log.httpStatus >= 200 && log.httpStatus < 300
            ? GREEN
            : log.httpStatus >= 400 && log.httpStatus < 500
              ? YELLOW
              : RED;

        const responseDetails = formatResponseDetails(log);
        const errorSuffix = log.errorMessage
          ? ` ${RED}${log.errorMessage}${RESET_COLOR}`
          : "";
        output += `${DIM}${log.timestamp.slice(11, 19)}${RESET_COLOR} ${GRAY}${sessionIdShort}${RESET_COLOR} ${log.serverName} → ${CYAN}${clientLabel}${RESET_COLOR} ${statusColor}(${log.httpStatus}, ${log.duration}ms)${RESET_COLOR}${responseDetails}${errorSuffix}\n`;
      }
    }
    output += "\n";
  }

  return output;
}

// Format request-specific details
function formatRequestDetails(log: LogEntry): string {
  if (!log.request?.params) return "";

  const params = log.request.params as Record<string, unknown>;

  // tools/call: show tool name and args
  if (log.method === "tools/call") {
    const toolName = params.name as string;
    const args = params.arguments as Record<string, unknown>;
    const argStr = Object.entries(args || {})
      .map(([k, v]) => `${k}:${JSON.stringify(v)}`)
      .join(", ");
    return ` ${DIM}${toolName}${argStr ? `(${argStr})` : ""}${RESET_COLOR}`;
  }

  // resources/read: show URI
  if (log.method === "resources/read") {
    const uri = params.uri as string;
    return ` ${DIM}${uri}${RESET_COLOR}`;
  }

  return "";
}

// Format response-specific details
function formatResponseDetails(log: LogEntry): string {
  if (!log.response || "error" in log.response) return "";

  const result = log.response.result as Record<string, unknown>;

  // initialize: show server name and version
  if (log.method === "initialize") {
    const serverInfo = result.serverInfo as { name: string; version: string };
    return ` ${DIM}→ ${serverInfo.name}@${serverInfo.version}${RESET_COLOR}`;
  }

  // tools/list: show count and first few tool names
  if (log.method === "tools/list") {
    const tools = result.tools as Array<{ name: string }>;
    const toolNames = tools.slice(0, 3).map((t) => t.name);
    const more = tools.length > 3 ? `, +${tools.length - 3}` : "";
    return ` ${DIM}${tools.length} tools: ${toolNames.join(", ")}${more}${RESET_COLOR}`;
  }

  // tools/call: show result text (truncated)
  if (log.method === "tools/call") {
    const content = result.content as Array<{ type: string; text?: string }>;
    const textContent = content?.find((c) => c.type === "text")?.text;
    if (textContent) {
      const truncated =
        textContent.length > 40
          ? `${textContent.slice(0, 40)}...`
          : textContent;
      return ` ${DIM}"${truncated}"${RESET_COLOR}`;
    }
  }

  // resources/list: show count
  if (log.method === "resources/list") {
    const resources = result.resources as unknown[];
    return ` ${DIM}${resources?.length || 0} resources${RESET_COLOR}`;
  }

  // prompts/list: show count
  if (log.method === "prompts/list") {
    const prompts = result.prompts as unknown[];
    return ` ${DIM}${prompts?.length || 0} prompts${RESET_COLOR}`;
  }

  return "";
}
