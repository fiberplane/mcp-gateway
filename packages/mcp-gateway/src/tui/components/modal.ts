import packageJson from "../../../package.json" with { type: "json" };
import type { Registry, ServerHealth } from "../../registry.js";
import type { DeleteServerState, FormState, ModalContent } from "../state.js";
import {
  BOLD,
  CYAN,
  DIM,
  GRAY,
  GREEN,
  RED,
  RESET_COLOR,
  YELLOW,
} from "./formatting.js";

function getHealthColor(health?: ServerHealth): string {
  switch (health) {
    case "up":
      return GREEN;
    case "down":
      return RED;
    default:
      return GRAY;
  }
}

function getHealthIndicator(health?: ServerHealth): string {
  const color = getHealthColor(health);
  return `${color}●${RESET_COLOR}`;
}

export function renderModal(
  content: ModalContent,
  formState?: FormState,
  deleteServerState?: DeleteServerState,
  registry?: Registry,
): string {
  let output = "";

  if (content === "mcp_instructions") {
    output += renderMcpInstructions();
    output += "\n";
    output += `${DIM}Press any key to go back${RESET_COLOR}\n`;
  } else if (content === "add_server_form" && formState) {
    output += renderAddServerForm(formState);
  } else if (
    content === "delete_server_form" &&
    deleteServerState &&
    registry
  ) {
    output += renderDeleteServerForm(deleteServerState, registry);
  }

  return output;
}

function renderMcpInstructions(): string {
  let output = "";
  output += `${CYAN}Fiberplane MCP Gateway v${packageJson.version}${RESET_COLOR}\n`;
  output += "\n";

  output += `${GREEN}This MCP Gateway is also an MCP server.${RESET_COLOR}\n`;
  output += `${YELLOW}Endpoint:${RESET_COLOR} http://localhost:3333/mcp\n`;
  output += `${DIM}Connect it to your agent (MCP Client) and use it to manage${RESET_COLOR}\n`;
  output += `${DIM}other MCP server connections, or inspect and analyze captured traffic${RESET_COLOR}\n`;
  output += "\n";

  // Available Tools
  output += `${GREEN}Available Tools:${RESET_COLOR}\n`;
  output += `  ${CYAN}•${RESET_COLOR} add_server      - Register new MCP servers\n`;
  output += `  ${CYAN}•${RESET_COLOR} remove_server   - Remove registered servers\n`;
  output += `  ${CYAN}•${RESET_COLOR} list_servers    - List all servers\n`;
  output += `  ${CYAN}•${RESET_COLOR} search_records  - Search captured traffic\n`;
  output += `  ${CYAN}•${RESET_COLOR} analyze_session - Analyze session flows\n`;
  output += `  ${CYAN}•${RESET_COLOR} get_server_stats - Server statistics\n`;
  output += "\n";

  // Example Config
  output += `${GREEN}Example MCP Client Config (Claude Desktop):${RESET_COLOR}\n`;
  output += `${DIM}{${RESET_COLOR}\n`;
  output += `${DIM}  "mcpServers": {${RESET_COLOR}\n`;
  output += `${DIM}    "mcp-gateway": {${RESET_COLOR}\n`;
  output += `${DIM}      "url": ${RESET_COLOR}${YELLOW}"http://localhost:3333/mcp"${RESET_COLOR}\n`;
  output += `${DIM}    }${RESET_COLOR}\n`;
  output += `${DIM}  }${RESET_COLOR}\n`;
  output += `${DIM}}${RESET_COLOR}\n`;
  output += "\n";

  // Storage
  output += `${GREEN}Storage:${RESET_COLOR}\n`;
  output += `${DIM}All captured traffic is stored in: ~/.mcp-gateway${RESET_COLOR}\n`;

  return output;
}

function renderAddServerForm(formState: FormState): string {
  let output = "";
  output += `${CYAN}Add New Server${RESET_COLOR}\n`;
  output += `${GRAY}${"─".repeat(60)}${RESET_COLOR}\n`;
  output += "\n";

  // Render each field
  for (let i = 0; i < formState.fields.length; i++) {
    const field = formState.fields[i];
    if (!field) continue;

    const isFocused = i === formState.focusedFieldIndex;

    // Label with hint text on same line
    const hint = field.placeholder
      ? ` ${DIM}(${field.placeholder})${RESET_COLOR}`
      : "";
    output += `${isFocused ? BOLD : DIM}${field.label}${hint}:${RESET_COLOR}\n`;

    // Input box with value and cursor
    const displayValue = field.value;
    const cursor = isFocused ? "_" : "";
    const prefix = isFocused ? `${YELLOW}>${RESET_COLOR} ` : "  ";
    output += `${prefix}${displayValue}${cursor}\n`;

    // Error message
    if (field.error) {
      output += `${RED}  ${field.error}${RESET_COLOR}\n`;
    }

    output += "\n";
  }

  output += `${GRAY}${"─".repeat(60)}${RESET_COLOR}\n`;
  output += `${DIM}[Tab] Next • [Enter] Submit • [ESC] Cancel${RESET_COLOR}\n`;

  return output;
}

function renderDeleteServerForm(
  deleteServerState: DeleteServerState,
  registry: Registry,
): string {
  let output = "";

  if (deleteServerState.showConfirm) {
    const server = registry.servers[deleteServerState.selectedIndex];
    if (!server) return output;

    const healthIndicator = getHealthIndicator(server.health);
    const nameColor = getHealthColor(server.health);

    output += `${CYAN}Confirm Deletion${RESET_COLOR}\n`;
    output += `${GRAY}${"─".repeat(60)}${RESET_COLOR}\n`;
    output += "\n";
    output += `${YELLOW}Really remove ${healthIndicator} ${nameColor}${server.name}${RESET_COLOR}?${RESET_COLOR}\n`;
    output += `${DIM}${server.url}${RESET_COLOR}\n`;
    output += "\n";
    output += `${DIM}Note: Capture history will be preserved on disk${RESET_COLOR}\n`;
    output += "\n";
    output += `${GRAY}${"─".repeat(60)}${RESET_COLOR}\n`;
    output += `${DIM}[Enter] Confirm • [ESC] Cancel${RESET_COLOR}\n`;
  } else {
    output += `${CYAN}Remove Server${RESET_COLOR}\n`;
    output += `${GRAY}${"─".repeat(60)}${RESET_COLOR}\n`;
    output += "\n";

    if (registry.servers.length === 0) {
      output += `${DIM}No servers to remove${RESET_COLOR}\n`;
      output += "\n";
      output += `${GRAY}${"─".repeat(60)}${RESET_COLOR}\n`;
      output += `${DIM}Press any key to go back${RESET_COLOR}\n`;
      return output;
    }

    output += `${DIM}Select a server to remove:${RESET_COLOR}\n`;
    output += "\n";

    for (let i = 0; i < registry.servers.length; i++) {
      const server = registry.servers[i];
      if (!server) continue;

      const isSelected = i === deleteServerState.selectedIndex;
      const marker = isSelected ? `${YELLOW}►${RESET_COLOR}` : " ";
      const healthIndicator = getHealthIndicator(server.health);
      const nameColor = isSelected ? YELLOW : getHealthColor(server.health);

      output += `${marker} ${healthIndicator} ${nameColor}${server.name}${RESET_COLOR} ${DIM}(${server.url})${RESET_COLOR}\n`;
    }

    output += "\n";
    output += `${GRAY}${"─".repeat(60)}${RESET_COLOR}\n`;
    output += `${DIM}[↑/↓] Navigate • [Enter] Select • [ESC] Cancel${RESET_COLOR}\n`;
  }

  return output;
}
