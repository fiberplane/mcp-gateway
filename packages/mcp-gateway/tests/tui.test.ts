import { expect, test } from "bun:test";
import type { Registry } from "../src/registry.js";
import { renderMenu } from "../src/tui/components/menu.js";
import { renderModal } from "../src/tui/components/modal.js";
import type { LogEntry } from "../src/tui/state.js";

// TUI Snapshot Tests
//
// These test the pure rendering functions (renderMenu, renderModal) without needing a TTY.
// Snapshots capture the exact output including ANSI escape codes for colors and formatting.
//
// Benefits:
// - Fast: No need to spawn processes or allocate PTYs
// - Reliable: Pure functions with deterministic output
// - Comprehensive: Can test all UI states easily
// - Regression detection: Catches unintended visual changes
//
// To update snapshots after intentional UI changes: bun test --update-snapshots
// For interactive testing with real TTY, run: bun run ./src/run.ts in tmux

test("TUI snapshot: empty menu with no servers", () => {
  const registry: Registry = { servers: [] };
  const logs: LogEntry[] = [];

  const output = renderMenu(registry, logs);

  expect(output).toMatchSnapshot();
});

test("TUI snapshot: menu with single HTTP server", () => {
  const registry: Registry = {
    servers: [
      {
        name: "test-server",
        url: "https://api.example.com/mcp",
        type: "http",
        headers: {},
        lastActivity: null,
        exchangeCount: 0,
      },
    ],
  };
  const logs: LogEntry[] = [];

  const output = renderMenu(registry, logs);

  expect(output).toMatchSnapshot();
});

test("TUI snapshot: menu with multiple servers and health status", () => {
  const registry: Registry = {
    servers: [
      {
        name: "healthy-server",
        url: "https://api1.example.com/mcp",
        type: "http",
        headers: {},
        lastActivity: new Date("2025-09-30T12:00:00Z").toISOString(),
        exchangeCount: 42,
        health: "up",
        lastHealthCheck: new Date("2025-09-30T12:00:00Z").toISOString(),
      },
      {
        name: "unhealthy-server",
        url: "https://api2.example.com/mcp",
        type: "http",
        headers: {},
        lastActivity: null,
        exchangeCount: 0,
        health: "down",
        lastHealthCheck: new Date("2025-09-30T12:00:00Z").toISOString(),
      },
      {
        name: "unknown-server",
        url: "https://api3.example.com/mcp",
        type: "http",
        headers: {},
        lastActivity: new Date("2025-09-30T11:30:00Z").toISOString(),
        exchangeCount: 15,
        health: "unknown",
      },
    ],
  };
  const logs: LogEntry[] = [];

  const output = renderMenu(registry, logs);

  expect(output).toMatchSnapshot();
});

test("TUI snapshot: menu with logs", () => {
  const registry: Registry = { servers: [] };
  const logs = [
    {
      timestamp: "2025-09-30T12:00:00.000Z",
      serverName: "test-server",
      sessionId: "abc12345-6789",
      method: "tools/list",
      httpStatus: 200,
      duration: 42,
      direction: "request" as const,
    },
    {
      timestamp: "2025-09-30T12:00:00.050Z",
      serverName: "test-server",
      sessionId: "abc12345-6789",
      method: "tools/list",
      httpStatus: 200,
      duration: 50,
      direction: "response" as const,
    },
    {
      timestamp: "2025-09-30T12:00:01.000Z",
      serverName: "test-server",
      sessionId: "def98765-4321",
      method: "tools/call",
      httpStatus: 500,
      duration: 100,
      direction: "response" as const,
      errorMessage: "Internal server error",
    },
  ];

  const output = renderMenu(registry, logs);

  expect(output).toMatchSnapshot();
});

test("TUI snapshot: add server form with empty fields", () => {
  const formState = {
    fields: [
      {
        name: "name",
        label: "Server Name",
        value: "",
        placeholder: "my-server",
      },
      {
        name: "url",
        label: "URL",
        value: "",
        placeholder: "https://api.example.com/mcp",
      },
    ],
    focusedFieldIndex: 0,
  };

  const output = renderModal("add_server_form", formState, undefined, {
    servers: [],
  });

  expect(output).toMatchSnapshot();
});

test("TUI snapshot: add server form with partial input", () => {
  const formState = {
    fields: [
      {
        name: "name",
        label: "Server Name",
        value: "my-api",
      },
      {
        name: "url",
        label: "URL",
        value: "https://api.example.com",
        error: "Invalid URL format",
      },
    ],
    focusedFieldIndex: 1,
  };

  const output = renderModal("add_server_form", formState, undefined, {
    servers: [],
  });

  expect(output).toMatchSnapshot();
});

test("TUI snapshot: delete server selection list", () => {
  const registry: Registry = {
    servers: [
      {
        name: "server1",
        url: "https://api1.example.com/mcp",
        type: "http",
        headers: {},
        lastActivity: null,
        exchangeCount: 0,
        health: "up",
      },
      {
        name: "server2",
        url: "https://api2.example.com/mcp",
        type: "http",
        headers: {},
        lastActivity: null,
        exchangeCount: 0,
        health: "down",
      },
    ],
  };

  const deleteServerState = {
    selectedIndex: 1,
    showConfirm: false,
  };

  const output = renderModal(
    "delete_server_form",
    undefined,
    deleteServerState,
    registry,
  );

  expect(output).toMatchSnapshot();
});

test("TUI snapshot: delete server confirmation", () => {
  const registry: Registry = {
    servers: [
      {
        name: "production-api",
        url: "https://prod.example.com/mcp",
        type: "http",
        headers: {},
        lastActivity: null,
        exchangeCount: 0,
        health: "down",
      },
    ],
  };

  const deleteServerState = {
    selectedIndex: 0,
    showConfirm: true,
  };

  const output = renderModal(
    "delete_server_form",
    undefined,
    deleteServerState,
    registry,
  );

  expect(output).toMatchSnapshot();
});

test("TUI snapshot: MCP instructions modal", () => {
  const output = renderModal("mcp_instructions", undefined, undefined, {
    servers: [],
  });

  expect(output).toMatchSnapshot();
});
