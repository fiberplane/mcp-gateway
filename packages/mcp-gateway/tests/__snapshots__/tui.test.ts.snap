// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`TUI snapshot: empty menu with no servers 1`] = `
"\x1B[96mFiberplane MCP Gateway v0.2.3\x1B[0m
\x1B[2mGateway: http://localhost:3333\x1B[0m
\x1B[2mMCP: http://localhost:3333/gateway/mcp\x1B[0m

\x1B[2mNo servers registered\x1B[0m

\x1B[93m[a]\x1B[0m Add server
\x1B[2m[d] Delete server\x1B[0m
\x1B[2m[c] Clear activity\x1B[0m
\x1B[93m[m]\x1B[0m MCP instructions
\x1B[93m[q]\x1B[0m Quit

"
`;

exports[`TUI snapshot: menu with single HTTP server 1`] = `
"\x1B[96mFiberplane MCP Gateway v0.2.3\x1B[0m
\x1B[2mGateway: http://localhost:3333\x1B[0m
\x1B[2mMCP: http://localhost:3333/gateway/mcp\x1B[0m

\x1B[96mServers:\x1B[0m

\x1B[38;5;245m●\x1B[0m \x1B[38;5;245mtest-server\x1B[0m
\x1B[96mhttp://localhost:3333/servers/test-server/mcp\x1B[0m \x1B[2m→ https://api.example.com/mcp\x1B[0m
\x1B[2mLast active: — • 0 exchanges\x1B[0m

\x1B[93m[a]\x1B[0m Add server
\x1B[93m[d]\x1B[0m Delete server
\x1B[2m[c] Clear activity\x1B[0m
\x1B[93m[m]\x1B[0m MCP instructions
\x1B[93m[q]\x1B[0m Quit

"
`;

exports[`TUI snapshot: menu with multiple servers and health status 1`] = `
"\x1B[96mFiberplane MCP Gateway v0.2.3\x1B[0m
\x1B[2mGateway: http://localhost:3333\x1B[0m
\x1B[2mMCP: http://localhost:3333/gateway/mcp\x1B[0m

\x1B[96mServers:\x1B[0m

\x1B[92m●\x1B[0m \x1B[92mhealthy-server\x1B[0m
\x1B[96mhttp://localhost:3333/servers/healthy-server/mcp\x1B[0m \x1B[2m→ https://api1.example.com/mcp\x1B[0m
\x1B[2mLast active: 3h ago • 42 exchanges\x1B[0m

\x1B[91m●\x1B[0m \x1B[91munhealthy-server\x1B[0m
\x1B[96mhttp://localhost:3333/servers/unhealthy-server/mcp\x1B[0m \x1B[2m→ https://api2.example.com/mcp\x1B[0m
\x1B[2mLast active: — • 0 exchanges\x1B[0m

\x1B[38;5;245m●\x1B[0m \x1B[38;5;245munknown-server\x1B[0m
\x1B[96mhttp://localhost:3333/servers/unknown-server/mcp\x1B[0m \x1B[2m→ https://api3.example.com/mcp\x1B[0m
\x1B[2mLast active: 4h ago • 15 exchanges\x1B[0m

\x1B[93m[a]\x1B[0m Add server
\x1B[93m[d]\x1B[0m Delete server
\x1B[2m[c] Clear activity\x1B[0m
\x1B[93m[m]\x1B[0m MCP instructions
\x1B[93m[q]\x1B[0m Quit

"
`;

exports[`TUI snapshot: menu with logs 1`] = `
"\x1B[96mFiberplane MCP Gateway v0.2.3\x1B[0m
\x1B[2mGateway: http://localhost:3333\x1B[0m
\x1B[2mMCP: http://localhost:3333/gateway/mcp\x1B[0m

\x1B[2mNo servers registered\x1B[0m

\x1B[93m[a]\x1B[0m Add server
\x1B[2m[d] Delete server\x1B[0m
\x1B[93m[c]\x1B[0m Clear activity
\x1B[93m[m]\x1B[0m MCP instructions
\x1B[93m[q]\x1B[0m Quit

\x1B[96mRecent Activity:\x1B[0m
\x1B[2m12:00:00\x1B[0m \x1B[38;5;245m[abc12345]\x1B[0m \x1B[96mclient\x1B[0m → test-server \x1B[2mtools/list\x1B[0m
\x1B[2m12:00:00\x1B[0m \x1B[38;5;245m[abc12345]\x1B[0m test-server → \x1B[96mclient\x1B[0m \x1B[92m(200, 50ms)\x1B[0m
\x1B[2m12:00:01\x1B[0m \x1B[38;5;245m[def98765]\x1B[0m test-server → \x1B[96mclient\x1B[0m \x1B[91m(500, 100ms)\x1B[0m \x1B[91mInternal server error\x1B[0m

"
`;

exports[`TUI snapshot: add server form with empty fields 1`] = `
"\x1B[96mAdd New Server\x1B[0m
\x1B[38;5;245m────────────────────────────────────────────────────────────\x1B[0m

\x1B[1mServer Name \x1B[2m(my-server)\x1B[0m:\x1B[0m
\x1B[93m>\x1B[0m _

\x1B[2mURL \x1B[2m(https://api.example.com/mcp)\x1B[0m:\x1B[0m
  

\x1B[38;5;245m────────────────────────────────────────────────────────────\x1B[0m
\x1B[2m[Tab] Next • [Enter] Submit • [ESC] Cancel\x1B[0m
"
`;

exports[`TUI snapshot: add server form with partial input 1`] = `
"\x1B[96mAdd New Server\x1B[0m
\x1B[38;5;245m────────────────────────────────────────────────────────────\x1B[0m

\x1B[2mServer Name:\x1B[0m
  my-api

\x1B[1mURL:\x1B[0m
\x1B[93m>\x1B[0m https://api.example.com_
\x1B[91m  Invalid URL format\x1B[0m

\x1B[38;5;245m────────────────────────────────────────────────────────────\x1B[0m
\x1B[2m[Tab] Next • [Enter] Submit • [ESC] Cancel\x1B[0m
"
`;

exports[`TUI snapshot: delete server selection list 1`] = `
"\x1B[96mRemove Server\x1B[0m
\x1B[38;5;245m────────────────────────────────────────────────────────────\x1B[0m

\x1B[2mSelect a server to remove:\x1B[0m

  \x1B[92m●\x1B[0m \x1B[92mserver1\x1B[0m \x1B[2m(https://api1.example.com/mcp)\x1B[0m
\x1B[93m►\x1B[0m \x1B[91m●\x1B[0m \x1B[93mserver2\x1B[0m \x1B[2m(https://api2.example.com/mcp)\x1B[0m

\x1B[38;5;245m────────────────────────────────────────────────────────────\x1B[0m
\x1B[2m[↑/↓] Navigate • [Enter] Select • [ESC] Cancel\x1B[0m
"
`;

exports[`TUI snapshot: delete server confirmation 1`] = `
"\x1B[96mConfirm Deletion\x1B[0m
\x1B[38;5;245m────────────────────────────────────────────────────────────\x1B[0m

\x1B[93mReally remove \x1B[91m●\x1B[0m \x1B[91mproduction-api\x1B[0m?\x1B[0m
\x1B[2mhttps://prod.example.com/mcp\x1B[0m

\x1B[2mNote: Capture history will be preserved on disk\x1B[0m

\x1B[38;5;245m────────────────────────────────────────────────────────────\x1B[0m
\x1B[2m[Enter] Confirm • [ESC] Cancel\x1B[0m
"
`;

exports[`TUI snapshot: MCP instructions modal 1`] = `
"\x1B[96mFiberplane MCP Gateway v0.2.3\x1B[0m

\x1B[92mThis MCP Gateway is also an MCP server.\x1B[0m
\x1B[93mEndpoint:\x1B[0m http://localhost:3333/gateway/mcp
\x1B[2mShort alias: http://localhost:3333/g/mcp\x1B[0m
\x1B[2mConnect it to your agent (MCP Client) and use it to manage\x1B[0m
\x1B[2mother MCP server connections, or inspect and analyze captured traffic\x1B[0m

\x1B[92mAvailable Tools:\x1B[0m
  \x1B[96m•\x1B[0m add_server      - Register new MCP servers
  \x1B[96m•\x1B[0m remove_server   - Remove registered servers
  \x1B[96m•\x1B[0m list_servers    - List all servers
  \x1B[96m•\x1B[0m search_records  - Search captured traffic

\x1B[92mExample MCP Client Config (Claude Desktop):\x1B[0m
\x1B[2m{\x1B[0m
\x1B[2m  "mcpServers": {\x1B[0m
\x1B[2m    "mcp-gateway": {\x1B[0m
\x1B[2m      "url": \x1B[0m\x1B[93m"http://localhost:3333/gateway/mcp"\x1B[0m
\x1B[2m    }\x1B[0m
\x1B[2m  }\x1B[0m
\x1B[2m}\x1B[0m

\x1B[92mStorage:\x1B[0m
\x1B[2mAll captured traffic is stored in: ~/.mcp-gateway\x1B[0m

\x1B[2mPress any key to go back\x1B[0m
"
`;
